### PR TITLE
Exclude sandbox runtime assets from session diffs

### PR DESCRIPTION
**Pull request by ColeMurray**
## Summary

- Track the exact Open-Inspect tools, skills, plugin files, and staged dependencies installed into a single-repository checkout.
- Write those paths to a marker-delimited block in the checkout-local Git `info/exclude` file.
- Preserve existing user exclude entries, unrelated `.opencode` files, and tracked changes.

## Why

Single-repository sessions run OpenCode from the repository itself. Sandbox startup installs Open-Inspect runtime assets under that checkout's `.opencode` directory after the session baseline is established. The Git-backed diff collector correctly included those untracked files, but they are runtime implementation details rather than user or agent changes.

Using `info/exclude` keeps the fix local to the sandbox checkout and avoids modifying the user's committed `.gitignore`. Exact owned paths are used instead of excluding all of `.opencode`, so user-authored configuration remains visible in session diffs.

This is stacked on #1036 because the exclusion behavior applies to its Git-backed session diff collector.

## Validation

- Sandbox runtime tests: 515 passed.
- Ruff checks passed for the changed runtime and test files.
- Tests cover preservation and idempotency, exact untracked-file visibility, tracked-file visibility, installed dependency ownership, bundled skills, and the optional Codex auth plugin.

## Summary by CodeRabbit

* **Bug Fixes**
  * Runtime-installed OpenCode assets are now hidden from local Git change detection using a managed `info/exclude` block, applied only when those runtime assets were actually installed.
  * Diff collection now filters out runtime-owned untracked changes while still surfacing user-authored updates and changes to already-tracked files.
  * Managed excludes preserve existing user entries and are idempotent across repeated sandbox setup runs.
* **Tests**
  * Expanded coverage for managed exclude persistence/idempotency, runtime ownership rules, asset-install tracking behavior, and diff/exclusion interactions (including override edge cases).

### CODE REVIEW

File: packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py [hunk 1]
```diff
@@ -750,15 +759,30 @@ def _stage_opencode_deps(deps_cache: Path, dest_dir: Path) -> None:
...
             shutil.copytree(cached_modules, local_modules, symlinks=True)
+            copied_modules = True
+        if copied_modules or (local_modules.is_dir() and "package.json" in installed):
+            installed.add("node_modules/")
+        if (
+            "package.json" in installed
+            and (deps_cache / "package-lock.json").is_file()
+            and (dest_dir / "package-lock.json").is_file()
+        ):
+            installed.add("package-lock.json")
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py]: Fixed in `c4f5cf9f`. Removed the fallback that claimed `package-lock.json` based on the neighboring `package.json`. The lockfile is now considered runtime-owned only when it was copied or its own contents match the staged lockfile. Added a regression test confirming a divergent existing user lockfile remains unchanged and outside the runtime-owned path set.

File: packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        raise ValueError(f"Runtime asset path must be repository-relative: {path!r}")
+    normalized = candidate.as_posix()
+    if normalized in ("", "."):
+        raise ValueError(f"Runtime asset path must name a file or directory: {path!r}")
+    return f"/{normalized}{'/' if trailing_slash else ''}"
+
+
+def install_runtime_git_excludes(repository: Path, runtime_paths: Iterable[str]) -> None:
+    """Replace our managed ``info/exclude`` block without touching user entries."""
+    patterns = sorted({_rooted_pattern(path) for path in runtime_paths})
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py]: Fixed in `1dcf7057`. Confirmed that a higher-precedence `.gitignore` negation can make the runtime asset appear in `git ls-files`. The managed block now serves as the persisted exact-path ownership manifest, and the diff collector filters those owned untracked paths after Git enumeration, so repository ignore rules cannot re-include them. Added a regression test that proves Git reports the negated path while the session capture still excludes it.

File: packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py [hunk 3]
```diff
@@ -750,15 +759,30 @@ def _stage_opencode_deps(deps_cache: Path, dest_dir: Path) -> None:
...
+                installed.add(name)
+            elif src.is_file() and dest.is_file() and filecmp.cmp(src, dest, shallow=False):
+                installed.add(name)
         cached_modules = deps_cache / "node_modules"
         local_modules = dest_dir / "node_modules"
+        copied_modules = False
         if cached_modules.is_dir() and not local_modules.exists():
             shutil.copytree(cached_modules, local_modules, symlinks=True)
+            copied_modules = True
+        if copied_modules or (local_modules.is_dir() and "package.json" in installed):
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py]: Fixed in `c542447b` (with the divergent lockfile portion handled in `c4f5cf9f`). A pre-existing `node_modules` tree is no longer claimed merely because `package.json` matches; new module-tree ownership is recorded only when this boot actually copies the tree. Added a regression test with a matching package manifest and user-created module tree. Cross-boot retention is being handled explicitly through the persisted ownership manifest in the separate snapshot-restoration thread.

File: packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        return
+
+    exclude_path = _git_exclude_path(repository)
+    try:
+        existing = exclude_path.read_text()
+    except FileNotFoundError:
+        existing = ""
+
+    block = "\n".join((BEGIN_MARKER, *patterns, END_MARKER)) + "\n"
+    start = existing.find(BEGIN_MARKER)
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py:101]: Fixed in `68e908c5`. The managed block is now treated as persisted ownership provenance: paths from prior boots remain owned while the corresponding checkout asset still exists, even when the current boot no longer installs it. Missing assets are pruned on reconciliation, and updates replace the manifest atomically. Added a multi-boot regression test covering a disabled retained tool and subsequent cleanup.

File: packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py [hunk 5]
```diff
@@ -750,15 +759,30 @@ def _stage_opencode_deps(deps_cache: Path, dest_dir: Path) -> None:
...
+            installed.add("node_modules/")
+        if (
+            "package.json" in installed
+            and (deps_cache / "package-lock.json").is_file()
+            and (dest_dir / "package-lock.json").is_file()
+        ):
+            installed.add("package-lock.json")
+        return installed
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/entrypoint.py]: Fixed in `c4f5cf9f` and `c542447b`. Each lockfile is now claimed only when it was copied or its own contents match the cache, and `node_modules` is newly claimed only when this boot copies that tree. Added separate regression tests for a divergent existing lockfile and a user-created module tree beside a matching package manifest.

File: packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py]: Fixed in `1dcf7057`. The managed block is read as the persisted exact-path ownership manifest, and the collector applies it after Git enumerates untracked paths. This gives runtime ownership final precedence independent of `.gitignore` negation. The regression test explicitly verifies that Git re-includes the path while session diff collection still excludes it.

File: packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    end = existing.find(END_MARKER, start + len(BEGIN_MARKER)) if start >= 0 else -1
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py]: Fixed in `68e908c5`. Reconciliation now retains previously recorded runtime paths while their checkout assets still exist, even if a later boot no longer installs them, and prunes ownership only after the asset disappears. The managed provenance is replaced atomically. Added a multi-boot shrinking-input regression test covering retained and subsequently removed assets.

File: packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        return frozenset()
+
+
+def is_runtime_git_excluded(path: str, runtime_paths: frozenset[str]) -> bool:
+    """Return whether a repository-relative path is covered by runtime ownership."""
+    return any(
+        path == runtime_path.rstrip("/")
+        or (runtime_path.endswith("/") and path.startswith(runtime_path))
+        for runtime_path in runtime_paths
+    )
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py:77]: Fixed in `baff45c6`. Runtime ownership now matches both the exact normalized path and descendants across a `/` boundary regardless of whether the persisted path has a trailing slash, while similarly prefixed siblings remain visible. Added a regression covering the exact directory, a descendant, and a sibling path; the full sandbox-runtime suite passes with 520 tests.

File: packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    return any(
+        path == runtime_path.rstrip("/") or path.startswith(runtime_path.rstrip("/") + "/")
+        for runtime_path in runtime_paths
+    )
+
+
+def _runtime_path_exists(repository: Path, path: str) -> bool:
+    candidate = repository / path.rstrip("/")
+    return candidate.exists() or candidate.is_symlink()
+
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py:83]: This is not applicable to the current invariant. `_runtime_path_exists()` is only called with values returned by `_managed_runtime_paths()`. That parser removes the single Git anchoring slash (`path = pattern[1:]`) and validates the resulting repository-relative value through `_rooted_pattern()`; `_rooted_pattern()` also rejects absolute inputs. Therefore `repository / path` cannot discard the repository base here. Adding `lstrip("/")` would duplicate normalization at a lower-level helper and could mask an invariant violation rather than expose it.

File: packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            return
+        separator = "" if not existing or existing.endswith(("\n", "\r")) else "\n"
+        updated = existing + separator + block
+
+    if updated == existing:
+        return
+    exclude_path.parent.mkdir(parents=True, exist_ok=True)
+    temporary_path = exclude_path.with_name(f".{exclude_path.name}.open-inspect.tmp")
+    temporary_path.write_text(updated)
+    temporary_path.replace(exclude_path)
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/git_excludes.py:121]: This concurrency scenario is outside the runtime lifecycle: the function has one call site, in the single sandbox supervisor during `start_opencode()`, before the OpenCode subprocess is spawned. A checkout is not initialized by multiple supervisor processes concurrently. The stable same-directory temporary name is overwritten on a later boot if a prior boot crashes, and `replace()` remains atomic. A UUID would only avoid the temporary-name collision; it would not make the preceding read/merge/write sequence concurrency-safe and would leave uniquely named residue on interruption. If concurrent writers become a supported lifecycle, this needs an explicit repository lock around the whole reconciliation rather than a unique filename.
